### PR TITLE
Mod select

### PR DESCRIPTION
**Pull request by DrabWeb**
Also adds `WaveOverlayContainer` to allow more overlays to use this animation

### CODE REVIEW

File: osu.Game/Modes/Mod.cs [hunk 1]
```diff
@@ -0,0 +1,428 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using System.ComponentModel;
+using osu.Game.Graphics;
+
+namespace osu.Game.Modes
+{
+    public abstract class Mod
```
Comment (Tom94) [inline osu.Game/Modes/Mod.cs]: Please xml-doc the public interface

File: osu.Game/Overlays/Mods/AssistedSection.cs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+using osu.Game.Graphics;
+using osu.Game.Modes;
+using osu.Game.Overlays.Mods;
+
+namespace osu.Game
+{
+    public class AssistedSection : ModSection
+    {
+        private ModButton relaxButton;
+        public ModButton RelaxButton
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/AssistedSection.cs]: All these should be `=>` one-liners

Comment (Tom94) [inline osu.Game/Overlays/Mods/AssistedSection.cs]: Also `{ get; private set; }` could make sense as an alternative.

File: osu.Game/Overlays/Mods/ModButton.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+using osu.Game.Graphics.Sprites;
+using osu.Game.Modes;
+using osu.Game.Modes.UI;
+
+namespace osu.Game.Overlays.Mods
+{
+    public class ModButton : FlowContainer
+    {
+        private ModIcon[] icons;
+        private ModIcon displayIcon
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/ModButton.cs]: =>

File: osu.Game/Overlays/Mods/ModButton.cs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                {
+                    if (Selected && i == icons.Length - 1) icons[i].Colour = SelectedColour;
+                    else icons[i].Colour = Colour;
+                }
+
+                displaySelectedMod();
+            }
+        }
+
+        public bool Selected
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/ModButton.cs]: =>

File: osu.Game/Overlays/Mods/ModButton.cs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                mods = value;
+                createIcons();
+                if (value.Length > 0)
+                {
+                    displayMod(value[0]);
+                }
+            }
+        }
+
+        public Mod SelectedMod
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/ModButton.cs]: =>, ?

File: osu.Game/Overlays/Mods/ModSection.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        public override bool IsPresent => true;
+    }
+
+    public class ModSection : Container
+    {
+        private OsuSpriteText headerLabel;
+
+        private AlwaysPresentFlowContainer buttonsContainer;
+        public FlowContainer ButtonsContainer
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/ModSection.cs]: =>

File: osu.Game/Overlays/Options/Sections/Graphics/LayoutOptions.cs [hunk 7]
```diff
@@ -21,7 +21,7 @@ private void load(FrameworkConfigManager config)
                 new OsuCheckbox
                 {
                     LabelText = "Fullscreen mode",
-                    Bindable = config.GetBindable<bool>(FrameworkConfig.Fullscreen),
+                    //Bindable = config.GetBindable<bool>(FrameworkConfig.Fullscreen),
```
Comment (Tom94) [inline osu.Game/Overlays/Options/Sections/Graphics/LayoutOptions.cs]: What happened here?

File: osu.Game/Overlays/Mods/ModSelectOverlay.cs [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+
+        private OsuSpriteText rankedLabel, multiplierLabel;
+        private FlowContainer rankedMultiplerContainer;
+
+        private FlowContainer modSectionsContainer;
+
+        private DifficultyReductionSection difficultyReductionSection;
+        private DifficultyIncreaseSection difficultyIncreaseSection;
+        private AssistedSection assistedSection;
+        private ModSection[] sections
```
Comment (Tom94) [inline osu.Game/Overlays/Mods/ModSelectOverlay.cs]: Creating a new array with every get seems weird. Either make the array a member and return it or make this a method `CreateSectionsArray`.

**Comment by peppy** [comment]
Mods shouldn't have any knowledge of play modes.

**Comment by peppy** [comment]
Going to merge this on the condition that it definitely requires further iterative refinement.

Good job getting it to this point 🚨.

**Pull request merged**